test/test_learn_detect.py

```diff
@@ -41,18 +71,80 @@ def callback(data):
+    if _SHOW:
+	cv2.imshow("Image window", cv_image)
+
+    if ( _INTERACTIVE ):
+        key = cv2.waitKey(10)
+    else:
+        cv2.waitKey(10)
+        key = 1048676
 
     if key == 1048684 or key == 108: # L
         print learn_srv(image=data, name=raw_input("Name? "))
     elif key == 1048676 or key == 100: # D
-        print detect_srv(image=data, external_api_request=external_api_request)
+        detections = detect_srv(image=data, external_api_request=external_api_request)
+        #print detections
+
+        if ( detections.face_detections.__len__() == 0):
+            disp_str = "(no detections)"
+            if ( camera_lr ):
+                disp_str = camera_lr + ": " + disp_str
+            print disp_str
+
+            if ( not camera_lr ):
+                msg_pub( '(no detections)' )
+            else:
+                msg_pub_with_pos( '(no detections)', -1, -1 )
+            return
+
+        #code.interact( local=locals() );
+
+        # unpack detections, find the winner
+        names = detections.face_detections[0].names
+        distances = detections.face_detections[0].l2_distances
+        c = detections.face_detections[0].x + detections.face_detections[0].width/2.
+        r = detections.face_detections[0].y + detections.face_detections[0].height/2.
+
+        min_distance = 1000.
+        min_distance_idx = 1000
+        for idx, val in enumerate( distances ):
+            if ( val < min_distance ):
+                min_distance = val
+                min_distance_idx = idx
+        #print distances
+        #print min_distance
+        #print min_distance_idx
+
+        if min_distance < _MAX_DISTANCE_FOR_MATCH:
+            #print min_distance_idx
+            match_name = names[ min_distance_idx ]
+            disp_str = names[ min_distance_idx ] + ' with distance ' + str( min_distance )
+            if ( camera_lr ):
+                disp_str = camera_lr + ": " + disp_str
+            print disp_str
+        else:
+            match_name = '(unknown)'
+            # occasionally, the first catch of no detections will not catch so make sure a distances really has something
+            if ( min_distance_idx != 1000 ):
+                disp_str ='(unknown)' + '; closest is ' + names[ min_distance_idx ] + ' with distance ' + str( min_distance ) 
+            else:
+                disp_str = '(unknown)'
+            if ( camera_lr ):
+                disp_str = camera_lr + ": " + disp_str
+            print disp_str
+
+        # publish
+        if ( not camera_lr ):
+            msg_pub( match_name )
+        else:
+            msg_pub_with_pos( match_name, r, c )
     elif key == 1048675 or key == 99: # C
         print clear_srv()
 
     return
 
+# subscribes to images with its callback (its callback learns/detects, calling the server)
 image_sub = rospy.Subscriber("image", Image, callback)
 rospy.loginfo("Listening to %s -- spinning .." % image_sub.name)
 rospy.loginfo("Usage: L to learn, D to detect")
```

Comment:
The "test" folder should be reserved for unit testing.  However since this was here before, I will allow it.